cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -779,6 +787,28 @@ Result<std::vector<int>> ParquetFileFragment::FilterRowGroups(
   return row_groups;
 }
 
+Result<std::vector<int>> ParquetFileFragment::FilterRangeRowGroups(
+    int64_t start_offset, int64_t length) {
+  std::vector<int> row_groups;
+  for (int row_group : *row_groups_) {
+    auto rg_metadata = metadata_->RowGroup(row_group);
+    std::shared_ptr<parquet::ColumnChunkMetaData> cc0 = rg_metadata->ColumnChunk(0);
+    int64_t r_start = cc0->data_page_offset();
+    if (cc0->has_dictionary_page() && r_start > cc0->dictionary_page_offset()) {
+      r_start = cc0->dictionary_page_offset();
+    }
+    int64_t r_bytes = 0L;
+    for (int col_id = 0; col_id < rg_metadata->num_columns(); col_id++) {
+      r_bytes += rg_metadata->ColumnChunk(col_id)->total_compressed_size();
```

Comment:
Could the begining of next rowgroup be used here?

Reply:
that feels a smart approach, however I'm not sure that all row groups are sorted in order.

Reply:
@zinking Would you mind find how other system split the rowgroups? `[start, offset]` might be tricky in some cases. RowGroups are usally in order, and here you require that column chunk is also ordered.

By the way, RowGroup might contain some metadata in the end of rowgroup. So I'm not sure do we need to considering this.

Reply:
you're referring to this: https://github.com/apache/parquet-mr/blob/d2c3c6d2e761a17b75d56fd356a37a2f754072f7/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java#L1287

Reply:
```
  /// This information is optional and may be 0 if omitted.
  int64_t total_compressed_size() const;
```

I think we can do same for cpp, if we assert on this value not 0

Reply:
* https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L836-L842 We cannot assure that, these two fields is optional
* Currently, rust implemention will not write these. See https://github.com/apache/arrow-rs/issues/4610

Reply:
`ColumnChunkMetadata.total_compressed_size` doesn't include the `ColumnChunkMetadata` thrift, which might be at end of ColumnChunk. Which might make the `length` be sightly smaller than expected.

Personally, I suggest to using other way to do this.